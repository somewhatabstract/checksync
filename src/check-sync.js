// @flow
import getMarkersFromFiles from "./get-markers-from-files.js";
import getFiles from "./get-files.js";
import Format from "./format.js";
import ErrorCodes from "./error-codes.js";
import handleViolations from "./handle-violations.js";
import cwdRelativePath from "./cwd-relative-path.js";

import type {ILog} from "./types.js";
import type {ErrorCode} from "./error-codes.js";

/**
 * Check the sync marks for the files represented by the given globs.
 *
 * @param {string[]} globs The globs that identify which files to check.
 * @param {boolean} autoFix When true, any out-of-date sync markers will be
 * updated.
 */
export default async function checkSync(
    globs: Array<string>,
    autoFix: boolean,
    comments: Array<string>,
    log: ILog,
): Promise<ErrorCode> {
    const files = await getFiles(globs);

    if (files.length === 0) {
        log.error("No matching files");
        return ErrorCodes.NO_FILES;
    }

    const cache = await getMarkersFromFiles(files, comments, log);

    if (log.errorsLogged && autoFix) {
        log.error(
            "Aborting fix due to parse errors. Fix errors and try again.",
        );
        return ErrorCodes.PARSE_ERRORS;
    }

    const violationFileNames = handleViolations(cache, autoFix, log).map(
        cwdRelativePath,
    );
    if (violationFileNames.length > 0) {
        if (autoFix) {
            // Output a summary of what we fixed.
            log.info(`Fixed ${violationFileNames.length} file(s)`);
        } else {
            // Output how to fix any violations we found if we're not running
            // autofix.
            const errorMsg = log.errorsLogged
                ? "Desynchronized blocks detected and parsing errors found. Fix the errors, update the blocks, then try:"
                : "Desynchronized blocks detected. Check them and update as required before resynchronizing:";
            log.group(Format.error(errorMsg));
            log.error(`checksync --fix ${violationFileNames.join(" ")}`, true);
            log.groupEnd();
            return ErrorCodes.DESYNCHRONIZED_BLOCKS;
        }
    }
    return ErrorCodes.SUCCESS;
}
