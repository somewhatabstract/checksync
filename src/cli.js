// @flow
/**
 * The implementation of our command line utility.
 */
import fs from "fs";
import chalk from "chalk";
import minimist from "minimist";
import checkSync from "./check-sync.js";
import Logger from "./logger.js";
import ErrorCodes from "./error-codes.js";
import logHelp from "./help.js";
import defaultArgs from "./default-args.js";
import parseGitIgnore from "parse-gitignore";

const ignoreFilesToExcludes = (ignoreFilesArg: string): Array<string> => {
    const ignoreFiles = ignoreFilesArg.split(",").filter(c => !!c);
    if (
        ignoreFilesArg === defaultArgs.ignoreFiles &&
        !fs.existsSync(ignoreFilesArg)
    ) {
        return [];
    }

    return ignoreFiles
        .map(file => fs.readFileSync(file))
        .map((content: Buffer) => parseGitIgnore(content))
        .reduce((prev, current: Array<string>) => [...prev, ...current], []);
};

/**
 * Run the command line.
 *
 * @param {string} launchFilePath
 */
export const run = (launchFilePath: string): void => {
    chalk.level = 3;

    const log = new Logger(console);

    // TODO(somewhatabstract): Verbose logging (make sure any caught errors
    // verbose their error messages)
    const args = minimist(process.argv, {
        boolean: ["updateTags", "dryRun", "help", "noIgnoreFile", "verbose"],
        string: ["comments", "rootMarker", "ignore", "ignoreFiles"],
        default: {
            ...defaultArgs,
        },
        alias: {
            comments: ["c"],
            dryRun: ["n", "dry-run"],
            help: ["h", "?"],
            ignore: ["i"],
            ignoreFiles: ["ignore-files"],
            noIgnoreFile: ["no-ignore-file"],
            rootMarker: ["m", "root-marker"],
            updateTags: ["u", "update-tags"],
        },
        unknown: arg => {
            // Filter out the node process.
            if (arg === process.execPath) return false;
            // Filter out our entry point.
            if (arg === launchFilePath) return false;
            // Filter out the command that yarn/npm might install.
            if (arg.endsWith(".bin/checksync")) return false;

            if (arg.startsWith("-")) {
                log.error(`Unknown argument: ${arg}`);
                process.exit(ErrorCodes.UNKNOWN_ARGS);
            }
            return true;
        },
    });

    if (args.help) {
        logHelp(log);
        process.exit(ErrorCodes.SUCCESS);
    }

    if (args.verbose) {
        log.setVerbose();
    }
    log.verbose(
        () => `Launched with args: ${JSON.stringify(args, null, "    ")}`,
    );

    const comments = ((args.comments: any): string).split(",").filter(c => !!c);
    const ignoreGlobs = ((args.ignore: any): string)
        .split(",")
        .filter(c => !!c);

    const ignoreFileGlobs = args.noIgnoreFile
        ? []
        : ignoreFilesToExcludes((args.ignoreFiles: any));
    const excludeGlobs = [...ignoreGlobs, ...ignoreFileGlobs];

    // Make sure we have something to search, so default to current working
    // directory if no globs are given.
    const includeGlobs = args._ && args._.length > 0 ? args._ : [process.cwd()];

    checkSync(
        {
            includeGlobs,
            excludeGlobs,
            autoFix: args.updateTags === true,
            comments,
            rootMarker: (args.rootMarker: any),
            dryRun: args.dryRun === true,
        },
        log,
    ).then(exitCode => {
        log.verbose(() => `Exiting with code ${exitCode}`);
        process.exit(exitCode);
    });
};
