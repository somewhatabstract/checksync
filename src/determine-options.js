// @flow
import findConfigurationFile from "./find-configuration-file.js";
import loadConfigurationFile from "./load-configuration-file.js";
import defaultOptions from "./default-options.js";

import type {ILog, Options} from "./types.js";
import type {minimistOutput} from "minimist";

export default async function determineOptions(
    args: minimistOutput,
    log: ILog,
): Promise<Options> {
    /**
     * The configuration we return here is a combination of the an RC file,
     * if found or wanted, and the supplied arguments.
     */
    if (args.config === false) {
        log.verbose(() => "Skipping configuration file search");
    }

    const configFilePath =
        args.config === false
            ? null
            : findConfigurationFile(
                  (args.config: any),
                  (args.rootMarker: any),
                  log,
              );

    const configFromFile =
        configFilePath == null
            ? null
            : await loadConfigurationFile(configFilePath, log);

    // We have to now build the options, with args taking precedence over
    // config. Then after that, put in any defaults we need.
    return {
        includeGlobs:
            (args._ && args._.length ? args._ : configFromFile?.includeGlobs) ??
            defaultOptions.includeGlobs,
        excludeGlobs:
            (args.ignore: any)?.split(";").filter((c) => !!c) ??
            configFromFile?.excludeGlobs ??
            defaultOptions.excludeGlobs,
        ignoreFiles:
            (args.ignoreFiles: any) === false
                ? []
                : (args.ignoreFiles: any)?.split(";").filter((c) => !!c) ??
                  configFromFile?.ignoreFiles ??
                  defaultOptions.ignoreFiles,
        autoFix:
            (args.updateTags: any) ??
            configFromFile?.autoFix ??
            defaultOptions.autoFix,
        json: (args.json: any) ?? configFromFile?.json ?? defaultOptions.json,
        comments:
            (args.comments: any)?.split(" ").filter((c) => !!c) ??
            configFromFile?.comments ??
            defaultOptions.comments,
        dryRun:
            (args.dryRun: any) ??
            configFromFile?.dryRun ??
            defaultOptions.dryRun,
        rootMarker:
            (args.rootMarker: any) ??
            configFromFile?.rootMarker ??
            defaultOptions.rootMarker,
    };
}
