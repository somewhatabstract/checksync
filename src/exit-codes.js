// @flow
/**
 * Exit codes that can be exited with.
 */
const exitCodes = {
    SUCCESS: 0,
    NO_FILES: 1,
    PARSE_ERRORS: 2,
    DESYNCHRONIZED_BLOCKS: 3,
    UNKNOWN_ARGS: 4,
    CATASTROPHIC: 5,
};

export type ExitCode = $Values<typeof exitCodes>;

export default exitCodes;
