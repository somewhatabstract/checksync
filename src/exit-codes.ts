/**
 * Exit codes that can be exited with.
 */
const exitCodes = {
    SUCCESS: 0,
    NO_FILES: 1,
    PARSE_ERRORS: 2,
    DESYNCHRONIZED_BLOCKS: 3,
    UNKNOWN_ARGS: 4,
    CATASTROPHIC: 5,
    BAD_CONFIG: 6,
} as const;

export type ExitCode = (typeof exitCodes)[keyof typeof exitCodes];

export default exitCodes;
