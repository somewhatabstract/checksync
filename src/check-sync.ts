import {ExitCode} from "./exit-codes";
import processCache from "./process-cache";

import {ILog, Options} from "./types";
import buildCache from "./build-cache";
import {loadCache} from "./load-cache";
import {ExitError} from "./exit-error";
import {outputCache} from "./output-cache";

/**
 *
 * Check the sync marks for the files represented by the given globs.
 *
 * @export
 * @param {Options} options The options for this run
 * @param {ILog} log A logger for outputting errors and the like.
 * @returns {Promise<ErrorCode>} The promise of an error code
 */
export default async function checkSync(
    options: Options,
    log: ILog,
): Promise<ExitCode> {
    // If the cacheMode is ignore, we just process the files normally.
    // If the cacheMode is write, then we build the cache but we don't
    // process it.
    // If the cacheMode is read, then we read the cache and process that.
    if (options.cacheMode !== "write" && options.autoFix && options.dryRun) {
        log.info("DRY-RUN: Files will not be modified");
    }

    try {
        const cache =
            options.cacheMode === "read"
                ? await loadCache(options.cachePath, log)
                : await buildCache(options, log);

        if (options.cacheMode === "write") {
            return outputCache(options, cache, log);
        }
        return processCache(options, cache, log);
    } catch (error) {
        if (error instanceof ExitError) {
            if (error.exitCode === ExitCode.BAD_CACHE) {
                log.error("Unable to load cache");
                return ExitCode.BAD_CACHE;
            }
            if (error.exitCode === ExitCode.NO_FILES) {
                log.error("No files found");
                return ExitCode.NO_FILES;
            }
        }
        log.error("Unexpected catastrophic error");
        return ExitCode.CATASTROPHIC;
    }
}
