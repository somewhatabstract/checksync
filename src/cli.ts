/**
 * The implementation of our command line utility.
 */
import fs from "fs";
import chalk from "chalk";
import checkSync from "./check-sync";
import Logger from "./logger";
import {ExitCode} from "./exit-codes";
import logHelp from "./help";
import {version} from "../package.json";
import determineOptions from "./determine-options";
import exit from "./exit";
import setCwd from "./set-cwd";
import {parseArgs} from "./parse-args";

/**
 * Run the command line.
 *
 * @param {string} launchFilePath
 */
export const run = async (launchFilePath: string): Promise<void> => {
    // Configure our logging output and argument parsing.
    chalk.level = 3;

    const log = new Logger(console);
    const args = await parseArgs(log);

    const isValidPath = (arg: string): boolean => {
        // Filter out the node process.
        if (arg === process.execPath) {
            return false;
        }
        // Filter out our entry point.
        if (arg === launchFilePath) {
            return false;
        }
        // Filter out the command that package manager might install.
        if (arg.endsWith(".bin/checksync")) {
            return false;
        }
        // Handle the entry point being a symlink
        try {
            const realpath = fs.realpathSync(arg);
            if (realpath === launchFilePath) {
                return false;
            }
        } catch {
            /* ignore errors, the arg may not be a path at all */
        }
        return true;
    };

    // We filter the set of file paths to verify them.
    const fileSet = new Set(args._);
    for (const file of fileSet) {
        if (!isValidPath(String(file))) {
            fileSet.delete(file);
        }
    }
    args._ = Array.from(fileSet);

    if (args.verbose) {
        log.setVerbose();
    }

    // Process arguments that fail early.
    if (args.version) {
        log.log(version);
        exit(log, ExitCode.SUCCESS);
    }

    if (args.help) {
        logHelp(log);
        exit(log, ExitCode.SUCCESS);
    }

    log.verbose(() => `Launched with args: ${JSON.stringify(args, null, 4)}`);

    // We need to apply the cwd argument to the process, otherwise the
    // configuration file discovery and other tasks will be working off a
    // different working directory.
    if (args.cwd != null) {
        setCwd(log, args.cwd);
    }

    // Parse arguments and configurations to get our options.
    return determineOptions(args, log)
        .then(
            (options) => checkSync(options, log),
            (e) => ExitCode.BAD_CONFIG,
        )
        .then(
            (exitCode) => {
                exit(log, exitCode);
            },
            (e) => {
                log.error(`Unexpected error: ${e}`);
                exit(log, ExitCode.CATASTROPHIC);
            },
        );
};
