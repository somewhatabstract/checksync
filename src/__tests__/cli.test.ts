import "jest-extended";
import * as minimist from "minimist";

import fs from "fs";
import {run} from "../cli";
import * as CheckSync from "../check-sync";
import {ExitCode} from "../exit-codes";
import Logger from "../logger";
import {version} from "../../package.json";
import * as DetermineOptions from "../determine-options";
import defaultOptions from "../default-options";
import * as Exit from "../exit";

jest.mock("minimist");
jest.mock("../logger", () => {
    const realLogger = jest.requireActual("../logger").default;
    const log = new realLogger(null);
    for (const key of Object.keys(log)) {
        if (typeof log[key] === "function") {
            jest.spyOn(log, key);
        }
    }
    return function () {
        return log;
    };
});
jest.mock("fs");
jest.mock("../../package.json", () => ({
    // We don't use the real version so that snapshots are consistent.
    version: "0.0.0",
}));

describe("#run", () => {
    it("should parse args", async () => {
        // Arrange
        const fakeParsedArgs: any = {
            fix: false,
            comments: "//,#",
            ignoreFile: undefined,
        } as const;
        const minimistSpy = jest
            .spyOn(minimist, "default")
            .mockReturnValue(fakeParsedArgs);
        jest.spyOn(DetermineOptions, "default").mockResolvedValue(
            defaultOptions,
        );
        jest.spyOn(CheckSync, "default").mockResolvedValue(0);
        // @ts-expect-error this is typed to never, but we want to mock it
        jest.spyOn(Exit, "default").mockImplementation(() => {});

        // Act
        await run(__filename);

        // Assert
        expect(minimistSpy).toHaveBeenCalledWith(
            process.argv,
            expect.any(Object),
        );
    });

    it.each(["help", "version"])(
        "should exit with success if %s arg present",
        (argName) => {
            // Arrange
            const fakeParsedArgs: any = {
                [argName]: true,
            } as const;
            jest.spyOn(minimist, "default").mockReturnValue(fakeParsedArgs);
            const exitSpy = jest
                .spyOn(Exit, "default")
                .mockImplementation(() => {
                    throw new Error("PRETEND PROCESS EXIT!");
                });

            // Act
            const underTest = () => run(__filename);

            // Assert
            expect(underTest).toThrowError("PRETEND PROCESS EXIT!");
            expect(exitSpy).toHaveBeenCalledWith(
                expect.anything(),
                ExitCode.SUCCESS,
            );
        },
    );

    it("should log version if version arg present", () => {
        // Arrange
        const fakeParsedArgs: any = {
            version: true,
        } as const;
        jest.spyOn(minimist, "default").mockReturnValue(fakeParsedArgs);
        jest.spyOn(Exit, "default").mockImplementation(() => {
            throw new Error("PRETEND PROCESS EXIT!");
        });
        const logSpy = jest.spyOn(new Logger(null), "log");

        // Act
        const underTest = () => run(__filename);

        // Assert
        expect(underTest).toThrow();
        expect(logSpy).toHaveBeenCalledWith(version);
    });

    it("should log help info if help arg present", () => {
        // Arrange
        const fakeParsedArgs: any = {
            help: true,
        } as const;
        jest.spyOn(minimist, "default").mockReturnValue(fakeParsedArgs);
        jest.spyOn(Exit, "default").mockImplementation(() => {
            throw new Error("PRETEND PROCESS EXIT!");
        });
        const logSpy = jest.spyOn(new Logger(null), "log");

        // Act
        const underTest = () => run(__filename);

        // Assert
        expect(underTest).toThrow();
        expect(logSpy).toHaveBeenCalledTimes(1);
        expect(logSpy.mock.calls[0][0]).toMatchSnapshot();
    });

    it("should change the working directory if cwd arg present", async () => {
        // Arrange
        const fakeParsedArgs: any = {
            cwd: "/some/path",
            verbose: true,
        } as const;
        jest.spyOn(minimist, "default").mockReturnValue(fakeParsedArgs);
        jest.spyOn(DetermineOptions, "default").mockResolvedValue(
            defaultOptions,
        );
        jest.spyOn(CheckSync, "default").mockResolvedValue(0);
        // @ts-expect-error this is typed to never, but we want to mock it
        jest.spyOn(Exit, "default").mockImplementation(() => {});
        const chdirSpy = jest
            .spyOn(process, "chdir")
            .mockImplementationOnce(() => {});

        // Act
        await run(__filename);

        // Assert
        expect(chdirSpy).toHaveBeenCalledWith("/some/path");
    });

    it("should exit with ExitCode.CATASTROPHIC if cwd arg present and chdir fails", async () => {
        // Arrange
        const fakeParsedArgs: any = {
            cwd: "/some/path",
        } as const;
        jest.spyOn(minimist, "default").mockReturnValue(fakeParsedArgs);
        jest.spyOn(DetermineOptions, "default").mockResolvedValue(
            defaultOptions,
        );
        jest.spyOn(CheckSync, "default").mockResolvedValue(0);
        const exitSpy = jest
            .spyOn(Exit, "default")
            // @ts-expect-error this is typed to never, but we want to mock it
            .mockImplementation(() => {});
        jest.spyOn(process, "chdir").mockImplementationOnce(() => {
            throw new Error("PRETEND CHDIR FAIL!");
        });

        // Act
        await run(__filename);

        // Assert
        expect(exitSpy).toHaveBeenCalledWith(
            expect.anything(),
            ExitCode.CATASTROPHIC,
        );
    });

    it("should pass arguments to determineOptions", async () => {
        // Arrange
        const fakeParsedArgs: any = {
            some: "args",
        } as const;
        jest.spyOn(minimist, "default").mockReturnValue(fakeParsedArgs);
        const determineOptionsSpy = jest
            .spyOn(DetermineOptions, "default")
            .mockResolvedValue(defaultOptions);
        jest.spyOn(CheckSync, "default").mockResolvedValue(0);
        // @ts-expect-error this is typed to never, but we want to mock it
        jest.spyOn(Exit, "default").mockImplementation(() => {});

        // Act
        await run(__filename);

        // Assert
        expect(determineOptionsSpy).toHaveBeenCalledWith(
            fakeParsedArgs,
            expect.any(Object),
        );
    });

    it("should invoke checkSync with determined options", async () => {
        // Arrange
        const fakeOptions = {
            ...defaultOptions,
            autoFix: true,
            comments: ["COMMENT1", "COMMENT2"],
            ignoreFiles: ["madeupfile"],
            excludeGlobs: ["glob1", "glob2"],
            includeGlobs: ["globs", "and globs"],
        } as const;
        jest.spyOn(minimist, "default").mockReturnValue({} as any);
        jest.spyOn(DetermineOptions, "default").mockResolvedValue(fakeOptions);
        const checkSyncSpy = jest
            .spyOn(CheckSync, "default")
            .mockResolvedValue(0);
        // @ts-expect-error this is typed to never, but we want to mock it
        jest.spyOn(Exit, "default").mockImplementation(() => {});

        // Act
        await run(__filename);

        // Assert
        expect(checkSyncSpy).toHaveBeenCalledWith(
            fakeOptions,
            expect.any(Object),
        );
    });

    it("should set logging to verbose when --verbose specified", async () => {
        const fakeParsedArgs: any = {
            verbose: true,
        } as const;
        jest.spyOn(CheckSync, "default").mockResolvedValue(0);
        jest.spyOn(DetermineOptions, "default").mockResolvedValue(
            defaultOptions,
        );
        jest.spyOn(minimist, "default").mockReturnValue(fakeParsedArgs);
        jest.spyOn(fs, "existsSync").mockReturnValueOnce(false);
        // @ts-expect-error this is typed to never, but we want to mock it
        jest.spyOn(Exit, "default").mockImplementation(() => {});
        const setVerboseSpy = jest.spyOn(new Logger(null), "setVerbose");

        // Act
        await run(__filename);

        // Assert
        expect(setVerboseSpy).toHaveBeenCalledTimes(1);
    });

    it("should exit process with the BAD_CONFIG code on determineOptions rejection", async () => {
        // Arrange
        jest.spyOn(DetermineOptions, "default").mockRejectedValue(
            new Error("Oh no, booms!"),
        );
        jest.spyOn(minimist, "default").mockReturnValue({} as any);
        const exitSpy = jest
            .spyOn(Exit, "default")
            // @ts-expect-error this is typed to never, but we want to mock it
            .mockImplementation(() => {});

        // Act
        await run(__filename);

        // Assert
        expect(exitSpy).toHaveBeenCalledWith(
            expect.anything(),
            ExitCode.BAD_CONFIG,
        );
    });

    it("should exit process with the exit code from checkSync", async () => {
        // Arrange
        jest.spyOn(CheckSync, "default").mockResolvedValue(42 as any);
        jest.spyOn(DetermineOptions, "default").mockResolvedValue(
            defaultOptions,
        );
        jest.spyOn(minimist, "default").mockReturnValue({} as any);
        const exitSpy = jest
            .spyOn(Exit, "default")
            // @ts-expect-error this is typed to never, but we want to mock it
            .mockImplementation(() => {});

        // Act
        await run(__filename);

        // Assert
        expect(exitSpy).toHaveBeenCalledWith(expect.anything(), 42);
    });

    it("should log error on rejection of checkSync method", async () => {
        // Arrange
        jest.spyOn(CheckSync, "default").mockRejectedValue(
            new Error("Oh no, booms!"),
        );
        jest.spyOn(DetermineOptions, "default").mockResolvedValue(
            defaultOptions,
        );
        jest.spyOn(minimist, "default").mockReturnValue({} as any);
        // @ts-expect-error this is typed to never, but we want to mock it
        jest.spyOn(Exit, "default").mockImplementation(() => {});
        const logSpy = jest.spyOn(new Logger(null), "error");

        // Act
        await run(__filename);

        // Assert
        expect(logSpy).toHaveBeenCalledWith(
            "Unexpected error: Error: Oh no, booms!",
        );
    });

    it("should exit with CATASTROPHIC code on rejection of checkSync method", async () => {
        // Arrange
        jest.spyOn(CheckSync, "default").mockRejectedValue(
            new Error("Oh no, booms!"),
        );
        jest.spyOn(DetermineOptions, "default").mockResolvedValue(
            defaultOptions,
        );
        jest.spyOn(minimist, "default").mockReturnValue({} as any);
        const exitSpy = jest
            .spyOn(Exit, "default")
            // @ts-expect-error this is typed to never, but we want to mock it
            .mockImplementation(() => {});

        // Act
        await run(__filename);

        // Assert
        expect(exitSpy).toHaveBeenCalledWith(
            expect.anything(),
            ExitCode.CATASTROPHIC,
        );
    });

    describe("unknown arg handling", () => {
        it("should return false for process.execPath", async () => {
            // Arrange
            jest.spyOn(CheckSync, "default").mockResolvedValue(0);
            jest.spyOn(DetermineOptions, "default").mockResolvedValue(
                defaultOptions,
            );
            // @ts-expect-error this is typed to never, but we want to mock it
            jest.spyOn(Exit, "default").mockImplementation(() => {});
            const minimistSpy = jest
                .spyOn(minimist, "default")
                .mockReturnValue({} as any);
            await run(__filename);
            const unknownHandler = minimistSpy.mock.calls[0][1]?.unknown;

            // Act
            const result = unknownHandler?.(process.execPath);

            // Assert
            expect(result).toBeFalse();
        });

        it("should return false for the given launchfile path", async () => {
            // Arrange
            jest.spyOn(CheckSync, "default").mockResolvedValue(0);
            jest.spyOn(DetermineOptions, "default").mockResolvedValue(
                defaultOptions,
            );
            // @ts-expect-error this is typed to never, but we want to mock it
            jest.spyOn(Exit, "default").mockImplementation(() => {});
            const minimistSpy = jest
                .spyOn(minimist, "default")
                .mockReturnValue({} as any);
            await run(__filename);
            const unknownHandler = minimistSpy.mock.calls[0][1]?.unknown;

            // Act
            const result = unknownHandler?.(__filename);

            // Assert
            expect(result).toBeFalse();
        });

        it("should return false for .bin command", async () => {
            // Arrange
            jest.spyOn(CheckSync, "default").mockResolvedValue(0);
            jest.spyOn(DetermineOptions, "default").mockResolvedValue(
                defaultOptions,
            );
            // @ts-expect-error this is typed to never, but we want to mock it
            jest.spyOn(Exit, "default").mockImplementation(() => {});
            const minimistSpy = jest
                .spyOn(minimist, "default")
                .mockReturnValue({} as any);
            await run(__filename);
            const unknownHandler = minimistSpy.mock.calls[0][1]?.unknown;

            // Act
            const result = unknownHandler?.("/Some/Path/To/.bin/checksync");

            // Assert
            expect(result).toBeFalse();
        });

        it("should return false for symlinked launch file", async () => {
            // Arrange
            jest.spyOn(CheckSync, "default").mockResolvedValue(0);
            jest.spyOn(DetermineOptions, "default").mockResolvedValue(
                defaultOptions,
            );
            // @ts-expect-error this is typed to never, but we want to mock it
            jest.spyOn(Exit, "default").mockImplementation(() => {});
            const minimistSpy = jest
                .spyOn(minimist, "default")
                .mockReturnValue({} as any);
            jest.spyOn(fs, "realpathSync").mockReturnValue(
                __filename, // Symlink resolves to our run file
            );
            await run(__filename);
            const unknownHandler = minimistSpy.mock.calls[0][1]?.unknown;

            // Act
            const result = unknownHandler?.("/usr/local/bin/checksync");

            // Assert
            expect(result).toBeFalse();
        });

        it("should exit on unknown arguments starting with -", async () => {
            // Arrange
            jest.spyOn(CheckSync, "default").mockResolvedValue(0);
            jest.spyOn(DetermineOptions, "default").mockResolvedValue(
                defaultOptions,
            );
            const exitSpy = jest
                .spyOn(Exit, "default")
                // @ts-expect-error this is typed to never, but we want to mock it
                .mockImplementation(() => {});
            const minimistSpy = jest
                .spyOn(minimist, "default")
                .mockReturnValue({} as any);
            await run(__filename);
            const unknownHandler = minimistSpy.mock.calls[0][1]?.unknown;

            // Act
            unknownHandler?.("--imadethisup");

            // Assert
            expect(exitSpy).toHaveBeenCalledWith(
                expect.anything(),
                ExitCode.UNKNOWN_ARGS,
            );
        });

        it("should report unknown arguments starting with -", async () => {
            // Arrange
            jest.spyOn(CheckSync, "default").mockResolvedValue(0);
            jest.spyOn(DetermineOptions, "default").mockResolvedValue(
                defaultOptions,
            );
            // @ts-expect-error this is typed to never, but we want to mock it
            jest.spyOn(Exit, "default").mockImplementation(() => {});
            const minimistSpy = jest
                .spyOn(minimist, "default")
                .mockReturnValue({} as any);
            const logSpy = jest.spyOn(new Logger(null), "error");
            await run(__filename);
            const unknownHandler = minimistSpy.mock.calls[0][1]?.unknown;

            // Act
            unknownHandler?.("--imadethisup");

            // Assert
            expect(logSpy).toHaveBeenCalledWith(
                "Unknown argument: --imadethisup",
            );
        });

        it("should return true for non-argument args (i.e. files)", async () => {
            // Arrange
            jest.spyOn(CheckSync, "default").mockResolvedValue(0);
            jest.spyOn(DetermineOptions, "default").mockResolvedValue(
                defaultOptions,
            );
            // @ts-expect-error this is typed to never, but we want to mock it
            jest.spyOn(Exit, "default").mockImplementation(() => {});
            const minimistSpy = jest
                .spyOn(minimist, "default")
                .mockReturnValue({} as any);
            await run(__filename);
            const unknownHandler = minimistSpy.mock.calls[0][1]?.unknown;

            // Act
            const result = unknownHandler?.("somethingelse");

            // Assert
            expect(result).toBeTrue();
        });
    });
});
