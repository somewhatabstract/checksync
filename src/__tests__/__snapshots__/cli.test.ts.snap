// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#run should log help info if help arg present 1`] = `
"[1m[32mchecksync 4.0.0 âœ… ðŸ”—[39m[22m

Checksync uses tags in your files to identify blocks that need to remain
synchronised. It works on any text file as long as it can find the tags.

[1m[32mTag Format[39m[22m

Each tagged block is identified by one or more sync-start tags and a single
sync-end tag.

The sync-start tags take the form:

    [1m[33m<comment> sync-start:<marker_id> <?checksum> <target_file>[39m[22m

The sync-end tags take the form:

    [1m[33m<comment> sync-end:<marker_id>[39m[22m

Each [1m[33mmarker_id[39m[22m can have multiple [1m[33msync-start[39m[22m tags, each with a different
target file, but there must be only one corresponding [1m[33msync-end[39m[22m tag.

Where:

    [1m[33m<comment>[39m[22m       is one of the comment tokens provided by the [1m[33m--comment[39m[22m
                    argument

    [1m[33m<marker_id>[39m[22m     is the unique identifier for this marker

    [1m[33m<checksum>[39m[22m      is the expected checksum of the corresponding block in
                    the target file

    [1m[33m<target_file>[39m[22m   is the path from your package root to the target file
                    with a corresponding sync block with the same [1m[33mmarker_id[39m[22m

[1m[32mUsage[39m[22m

[1m[33mchecksync <arguments> <include_paths>[39m[22m

Where:

    [1m[33m<arguments>[39m[22m        are the arguments you provide (see below).

    [1m[33m<include_paths>[39m[22m    are space-separated paths and glob patterns
                       for identifying files to check.
                       Defaults to all files below the current working
                       directory.

[1m[32mArguments[39m[22m

    [1m[33m--comments,-c[39m[22m      A string containing space-separated tokens that
                       indicate the start of lines where tags appear.
                       Defaults to [1m[33m"// #"[39m[22m.

    [1m[33m--config[39m[22m           Path to a JSON file containing configuration options.
                       When not specified, checksync will look for a config
                       file; use [1m[33m--no-config[39m[22m to disable this search.

                       The search starts relative to the current working
                       directory, first by looking adjacent to a marker file
                       match, otherwise, by looking for a configuration file in
                       or above the current working directory. Filenames that
                       are considered:
                           [1m[33m.checksyncrc[39m[22m
                           [1m[33m.checksyncrc.json[39m[22m

    [1m[33m--dry-run,-n[39m[22m       Ignored unless supplied with [1m[33m--update-tags[39m[22m.

    [1m[33m--help,-h[39m[22m          Outputs this help text.

    [1m[33m--ignore,-i[39m[22m        A string containing semi-colon-separated globs that
                       identify files that should not be checked.

    [1m[33m--ignore-files[39m[22m     A semi-colon-separated list of paths and globs that
                       identify .gitignore-format files defining patterns for
                       paths to be ignored. These will be combined with the
                       explicit [1m[33m--ignore[39m[22m globs.
                       Ignored if [1m[33m--no-ignore-file[39m[22m is present.
                       Defaults to [1m[33m.gitignore[39m[22m.

    [1m[33m--json,-j[39m[22m          Output errors and violations as JSON.

    [1m[33m--no-config[39m[22m        Prevents searching for a configuration file.
                       Ignored if [1m[33m--config[39m[22m is supplied.

    [1m[33m--no-ignore-files[39m[22m  When [1m[33mtrue[39m[22m, does not use any ignore file. This is
                       useful when the default value for [1m[33m--ignore-file[39m[22m is
                       not wanted.

    [1m[33m--root-marker,-m[39m[22m   By default, the root directory (used to generate
                       interpret and generate target paths for sync-start
                       tags) for your project is determined by the nearest
                       ancestor directory to the processed files that
                       contains a [1m[33mpackage.json[39m[22m file. If you want to
                       use a different file or directory to identify your
                       root directory, specify that using this argument.
                       For example, [1m[33m--root-marker .gitignore[39m[22m would mean
                       the first ancestor directory containing a
                       [1m[33m.gitignore[39m[22m file.

    [1m[33m--update-tags,-u[39m[22m   Updates tags with incorrect target checksums. This
                       modifies files in place; run with [1m[33m--dry-run[39m[22m to see
                       what files will change without modifying them.

    [1m[33m--verbose[39m[22m          More details will be added to the output when this
                       option is provided. This is useful when determining if
                       provided glob patterns are applying as expected, for
                       example.

    [1m[33m--version[39m[22m          Outputs the version and exits.

[1m[32mConfiguration Format[39m[22m
A configuration file is a JSON file containing configuration options. All
of the values are optional (defaults apply per the corresponding CLI arguments).
Arguments supplied along with a configuration file will override the
configuration file.

    [1m[33mautoFix[39m[22m             Equivalent to using the [1m[33m--update-tags[39m[22m option.

    [1m[33mdryRun[39m[22m              Equivalent to using the [1m[33m--dry-run[39m[22m option.

    [1m[33mcomments[39m[22m            Equivalent to using the [1m[33m--comments[39m[22m option,
                        except an array instead of space-separated string.

    [1m[33mignoreFiles[39m[22m         Equivalent to using the [1m[33m--ignore-files[39m[22m option,
                        except an array instead of semi-colon-separated string.
                        An empty array is equivalent to [1m[33m--no-ignore-files[39m[22m.

    [1m[33mexcludeGlobs[39m[22m        Equivalent to using the [1m[33m--ignore[39m[22m option, except
                        an array instead of semi-colon-separated string.

    [1m[33mincludeGlobs[39m[22m        Equivalent to the [1m[33minclude_paths[39m[22m passed at the
                        end of the command line.

    [1m[33mrootMarker[39m[22m          Equivalent to using the [1m[33m--root-marker[39m[22m option.

Example:
    [1m[33m{[39m[22m
[1m[33m        "autoFix": true,[39m[22m
[1m[33m        "dryRun": false,[39m[22m
[1m[33m        "comments": ["//", "#"],[39m[22m
[1m[33m        "ignoreFiles": [".gitignore"],[39m[22m
[1m[33m        "includeGlobs": ["**/*.js"],[39m[22m
[1m[33m        "excludeGlobs": ["**/node_modules/**"],[39m[22m
[1m[33m        "rootMarker": ".gitignore",[39m[22m
[1m[33m        "json": false[39m[22m
[1m[33m    }[39m[22m
"
`;
